02-arm-perf-report/filaments_32p_1n_1t_2021-01-22_16-55.txt

```diff
@@ -0,0 +1,63 @@
+Command:        /cosma/home/do008/dc-park1/STORM/storm3d/filaments
+Resources:      1 node (32 physical, 64 logical cores per node)
+Memory:         503 GiB per node
+Tasks:          32 processes
+Machine:        b109.pri.cosma7.alces.network
+Start time:     Fri Jan 22 16:55:58 2021
+Total time:     294 seconds (about 5 minutes)
+Full path:      /cosma/home/do008/dc-park1/STORM/storm3d
+
+Summary: filaments is Compute-bound in this configuration
+Compute:                                     68.3% |======|
+MPI:                                         30.6% |==|
+I/O:                                          1.1% ||
+This application run was Compute-bound. A breakdown of this time and advice for investigating further is in the CPU section below. 
+
+CPU:
+A breakdown of the 68.3% CPU time:
+Scalar numeric ops:                          34.9% |==|
+Vector numeric ops:                           0.7% ||
+Memory accesses:                             64.4% |=====|
```

Comment:
This looks quite worrying. Do we believe these figures?

Reply:
Maybe relevant - the Intel compilers were significantly worse for us when we checked on Marconi a couple of years ago, even though Marconi's an Intel machine and they're the recommended set. Vague guess is maybe they don't handle some templates or C++11 features well, maybe that stops them vectorizing properly? I can have another crack at using `perf-report` on the gnu-compiled version. @JosephThomasParker did `perf-report` work for you in the batch queue? It failed for me, and I'm wondering whether to just run on the login node.

Reply:
What was the error? It ran fine for me, and I'm not sure I'd want to run it on the login nodes, since it's a wrapper to a BOUT++ execution.

It would be good to see how gnu goes. I got a _lot_ of vectorization warning when compiling:
```
remark #11074: Inlining inhibited by limit max-size 
remark #11074: Inlining inhibited by limit max-total-size
```
I turned on the optimization reports (by putting `-qopt-report=4 -qopt-report-phase ipo` into `BOUT_FLAGS`). Intel reports are rather verbose though. Each source file has its own; `field3d.cxx` report is 24203 lines long. I've attached it for reference, but I need to remind myself how these things work. 
[field3d.optrpt.txt](https://github.com/boutproject/cs-performance-tuning-workshop/files/5868328/field3d.optrpt.txt)

Reply:
Below is the error message that I got from `perf-report` running in the batch queue. I gave up at that point...
```
Performance Reports: Processes 36,50,63: 
Performance Reports: 
Performance Reports: One of the debuggers Arm Performance Reports is communicating with has terminated unexpectedly.
Performance Reports: 
Performance Reports: The debugger exited with the following message:
Performance Reports: 
Performance Reports: &"info frame 8\n"
Performance Reports: &"/var/lib/jenkins/workspace/gdb/development/gdb-7.12.1/build-gdb-7.12.1-x86-64-artifacts/.conan/data/GDB_7_12_1/I1f7586c300e9096f9eb441629170cc8d5b834c9e.1/gdb/review/source/gdb/value.c:4361: internal-error: void value_fetch_lazy(value*): Assertion `frame != NULL' failed.\nA problem internal to GDB has been detected,\...
Performance Reports: 
Performance Reports: &"info frame 8\n"
Performance Reports: &"/var/lib/jenkins/workspace/gdb/development/gdb-7.12.1/build-gdb-7.12.1-x86-64-artifacts/.conan/data/GDB_7_12_1/I1f7586c300e9096f9eb441629170cc8d5b834c9e.1/gdb/review/source/gdb/value.c:4361: internal-error: void value_fetch_lazy(value*): Assertion `frame != NULL' failed.\nA problem internal to GDB has been detected,\nfurther debugging may prove unreliable.\n"
Performance Reports: &"\nThis is a bug, please report it."
Performance Reports: &"  For instructions, see:\n<http://www.gnu.org/software/gdb/bugs/>."
Performance Reports: &"\n\n"
Performance Reports: (gdb-allinea) 
Performance Reports: Aborted
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121() [0x60a1f2]
Performance Reports: /lib64/libc.so.6(+0x36400) [0x2af3e1ca6400]
Performance Reports: /lib64/libc.so.6(gsignal+0x37) [0x2af3e1ca6387]
Performance Reports: /lib64/libc.so.6(abort+0x148) [0x2af3e1ca7a78]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121() [0x6dd516]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121() [0x6df6fb]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121(_Z15internal_verrorPKciS0_P13__va_list_tag+0x19) [0x6df799]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121() [0x716a22]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121(_Z16value_fetch_lazyP5value+0x4bc) [0x5acadc]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121(_Z19value_contents_copyP5valuelS0_ll+0x35) [0x5ace65]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121(_Z25read_frame_register_valueP5valueP10frame_info+0xf0) [0x59db40]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121(_Z21address_from_registeriP10frame_info+0x108) [0x59e5c8]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121() [0x69058b]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121() [0x6909c3]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121() [0x6e2e37]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121(_Z21get_prev_frame_alwaysP10frame_info+0x190) [0x6e4e40]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121(_Z14get_prev_frameP10frame_info+0x238) [0x6e5798]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121() [0x600869]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121(_Z8cmd_funcP16cmd_list_elementPci+0x53) [0x528aa3]
Performance Reports: /cosma/local/arm/ddt/20.2.1/libexec/gdb-7121(_Z15execute_commandPci+0x253) [0x6db6a3]
Performance Reports:
```

Reply:
I posted the error message in the Slack #helpdesk channel, but nobody replied yet. Checking again it doesn't seem to be login node vs batch queue - it seems to be related to number of processes I use. 1, 2 or 4 is fine, 6 is flaky and 8 fails. I've put a follow-up with that info on Slack, maybe that'll help prompt a reply...

I ran a 4-process job, and got this report for test-isothermal
```
[dc-omot1@b101 test-isothermal]$ cat filaments_4p_1n_2021-01-25_19-02.txt 
Command:        mpirun -np 4 ../../storm3d/filaments
Resources:      1 node (32 physical, 64 logical cores per node)
Memory:         503 GiB per node
Tasks:          4 processes
Machine:        b102.pri.cosma7.alces.network
Start time:     Mon Jan 25 19:02:02 2021
Total time:     444 seconds (about 7 minutes)
Full path:      /cosma5/data/do008/dc-omot1/storm/storm3d

Summary: filaments is Compute-bound in this configuration
Compute:                                     93.8% |========|
MPI:                                          6.0% ||
I/O:                                          0.3% ||
This application run was Compute-bound. A breakdown of this time and advice for investigating further is in the CPU section below. 
As very little time is spent in MPI calls, this code may also benefit from running at larger scales.

CPU:
A breakdown of the 93.8% CPU time:
Scalar numeric ops:                          17.1% |=|
Vector numeric ops:                          10.2% ||
Memory accesses:                             61.0% |=====|
The per-core performance is memory-bound. Use a profiler to identify time-consuming loops and check their cache performance.
Little time is spent in vectorized instructions. Check the compiler's vectorization advice to see why key loops could not be vectorized.

MPI:
A breakdown of the 6.0% MPI time:
Time in collective calls:                     0.8% ||
Time in point-to-point calls:                99.2% |=========|
Effective process collective rate:             125 bytes/s
Effective process point-to-point rate:         508 MB/s
Most of the time is spent in point-to-point calls with an average transfer rate. Using larger messages and overlapping communication and computation may increase the effective transfer rate.

I/O:
A breakdown of the 0.3% I/O time:
Time in reads:                                0.0% |
Time in writes:                             100.0% |=========|
Effective process read rate:                  0.00 bytes/s
Effective process write rate:                 23.7 MB/s
Most of the time is spent in write operations with a low effective transfer rate. This may be caused by contention for the filesystem or inefficient access patterns. Use an I/O profiler to investigate which write calls are affected.

Threads:
A breakdown of how multiple threads were used:
Computation:                                 98.9% |=========|
Synchronization:                              1.1% ||
Physical core utilization:                   18.1% |=|
System load:                                 12.5% ||
Physical core utilization is low. Try increasing the number of threads or processes to improve performance.

Memory:
Per-process memory usage may also affect scaling:
Mean process memory usage:                     254 MiB
Peak process memory usage:                     266 MiB
Peak node memory usage:                       1.0% ||
The peak node memory usage is very low. Running with fewer MPI processes and more data on each process may be more efficient.

Energy:
A breakdown of how energy was used:
CPU:                                      not supported
System:                                   not supported
Mean node power:                          not supported
Peak node power:                              0.00 W
Energy metrics are not available on this system.
CPU metrics are not supported (no intel_rapl module)
```
So the vectorisation seems a bit better than Intel... Joseph, could you run a 4-process job to compare?

Reply:
Found this https://community.intel.com/t5/Intel-C-Compiler/Cause-for-remark-11074-Inlining-inhibited-by-limit-max-size/td-p/1073635 which suggested using `-no-inline-max-size -no-inline-max-total-size`. I'm trying to compile with icc/intelmpi on Marconi. I added those flags to `CXXFLAGS` and edited `make.config` to set `--std=c++14` (because the Intel compiler complained about things in v4.4.0-alpha without it). Still lots of warnings but the inlining ones seem to be gone.

Reply:
I tried compiling on marconi using `-no-inline-max-size -no-inline-max-total-size`. I had to submit the compilation to the debug queue because it took so long, and then running in the queue on 48 cores (with `make -j 48`) took about 40 mins to compile. I assume that was just one file (likely `index_derivs.cxx`?) that took ages... I also saw the compiler's memory usage going up to ~20GB, so running 48 jobs at once may not have been optimal. Once it did compile, I get an 'Illegal instruction' error when I try to run on the login node, but that seems to go away when a job runs on a compute node. I guess the problem is that Marconi's login nodes are Broadwell and compute nodes are Skylake, and I guess something in the modules (I have `env-skl/1.0` loaded) is giving me a compiler flag that produces code that's not compatible with the Broadwell processors. Weirdly I get the error on the login nodes even when configuring with `--enable-debug`, so it's happening with `-O0`!

I tried running test-isothermal compiled without `-no-inline-max-size -no-inline-max-total-size` on 1 node and run time is about the same as with gcc/openmpi. Am waiting now for storm to compile with those flags to compare...

Reply:
compiling with `-no-inline-max-size -no-inline-max-total-size` ran successfully, but shaved maybe a few seconds off the run time, 1m58s vs 2m04s, although I only ran each case once, so no statistics...

I tried running that test on various numbers of nodes (in brackets first nxpe, then the size of the grid on each processor):
1 node (nxpe=24, 8x8) : 131.11s
2 node (nxpe=24, 8x4) : 84.64s
4 node (nxpe=24, 8x2) : 64.48s
4 node (nxpe=48, 4x4) : 64.70s
8 node (nxpe=48, 4x2) : 81.30s
8 node (nxpe=96, 2x4) : 60.87s

Have to say I'm surprised - I'd expected that keeping the Laplace solve on one node (nxpe<=48) or even one NUMA region (nxpe<=24) would give better performance, but in the 8-node case nxpe=96 was faster than nxpe=48.

I feel as though the communication matrix plots that @d7919 suggested making are telling us something, but I haven't got my head around what they show. Is the pair of slow off-diagonals the set of processors on node boundaries maybe? I can upload the aps_report.html and aps_report_r2r_comm_matrix.html files if it's useful?

Sim info: nx=196, ny=16, nz=128. Slab grid with target plate y-boundaries at both ends. Marconi nodes are 48 procs.

Communication matrices
---------------------------------
1 node, nxpe=24
![1node](https://user-images.githubusercontent.com/3958036/106067337-69e88700-60f6-11eb-98ee-489b705bdc0c.png)

2 node, nxpe=24
![2node](https://user-images.githubusercontent.com/3958036/106067353-7076fe80-60f6-11eb-92fc-811761510de9.png)

4 node, nxpe=24
![4node-nxpe24](https://user-images.githubusercontent.com/3958036/106067390-7d93ed80-60f6-11eb-9f05-6a71c821cc0d.png)

4 node, nxpe=48
![4node-nxpe48](https://user-images.githubusercontent.com/3958036/106067409-85ec2880-60f6-11eb-845b-53a1e7feb23f.png)

8 node, nxpe=48
![8node-nxpe48](https://user-images.githubusercontent.com/3958036/106067428-8be20980-60f6-11eb-8023-2b8dc3ef2408.png)

8 node, nxpe=96
![8node-nxpe96](https://user-images.githubusercontent.com/3958036/106067440-93a1ae00-60f6-11eb-91e4-44a305de35ee.png)

Reply:
I'm coming at this before I'm fully awake/caffeinated, but at least in the first case with `nxpe=24`, are the off-diagonals processors adjacent in y but split across NUMA regions? That would explain why they both have a lot of communication and why it's slow, as the MPI messages can't be done in-memory.

Reply:
Interesting thought. I tried a run with nxpe=12 to see if I could see a difference between NUMA-split and -non-split processes. I don't think that's what this shows. Instead comms become asymmetric (and significantly slower than the nxpe24 case - the color scale is up to 11s here rather than 1.78s). Am looking at the docs on processor pinning now to try and figure out what NUMA regions the processes are actually being put into.

![1node-nxpe12](https://user-images.githubusercontent.com/3958036/106125658-4f44fb00-6154-11eb-978b-05118d72221b.png)

Reply:
I found the intelmpi process pinning docs https://software.intel.com/content/www/us/en/develop/documentation/mpi-developer-reference-linux/top/environment-variable-reference/process-pinning/environment-variables-for-process-pinning.html but setting `export I_MPI_DEBUG=4` shows the default does what we'd expect - each rank is assigned to the cpu with the same number. I assume that means that 0-23 are in one NUMA region and 24-47 are in the other.

Reply:
Actually, those red squares on the 1-node, nxpe=12 case are where the y-communication goes across the NUMA boundary. They're 36->24, 37->25, ..., 47->35.

Reply:
> compiling with `-no-inline-max-size -no-inline-max-total-size` ran successfully, but shaved maybe a few seconds off the run time, 1m58s vs 2m04s, although I only ran each case once, so no statistics...

@johnomotani did you look at the APS report with these flags on?  Did you get a higher percentage for "vector numeric ops"?

Reply:
I don't seem to get the `vector numeric ops` report on Marconi. Is there an option or flag that needs setting. I guess it might be hardware settings or Intel 2018 compilers...

Reply:
> I don't seem to get the `vector numeric ops` report on Marconi. Is there an option or flag that needs setting. I guess it might be hardware settings or Intel 2018 compilers...

I think it's reliant on access to hardware counters which aren't always enabled by default.

Reply:
I could ask the helpdesk to enable them (worst that can happen is they say no!) - do you know what exactly I'd need to ask them to do?

Reply:
I'm not sure that I'd know what to ask for beyond providing access to hardware counters / asking for advice as to why that information isn't appearing. It's possible that just loading the papi module (https://www.hpc.cineca.it/software/papi-0) is enough to give you this access. I'm well beyond my area of expertise now though!